src/SceneGraphDebugCommandController.ts

```diff
@@ -0,0 +1,305 @@
+    private echoLines = 0;
+    private timeout = 5000;
+    private port = 8080;
+
+    public async connect() {
+        this.removeConnection();
+
+        try {
+            // Make a new telnet connections object
+            let connection = new Telnet();
+
+            connection.on('close', this.removeConnection);
+
+            await connection.connect({
+                host: this.host,
+                port: this.port,
+                shellPrompt: this.shellPrompt,
+                echoLines: this.echoLines,
+                timeout: this.timeout
+            });
+            this.connection = connection;
+        } catch (error) {
+            throw new Error(error.message);
+        }
+    }
+
+    private removeConnection() {
+        this.connection = null;
+    }
+
+    /**
+     * executes the different bsprof commands used for brightscript profiling.
+     * @param {('pause'|'resume'|'status')} option Pause, resume, or get BS profiling status.
+     */
+    public async bsprof(option: 'pause'|'resume'|'status'): Promise<SceneGraphCommandResponse> {
+        return this.exec(`bsprof-${option}`);
+    }
+
+    /**
+     * Prints the current memory and CPU utilization of a channel (RAM usage is reported in KibiBytes [KiB]). The channel manifest must include the run_as_process=1 attribute to use this command.
+     *
+     * If an interval is provided the device repeats the command the specified number of seconds and outputs the results to port 8085.
+     * (Available since Roku OS 10.0)
+     * @param {{ interval: number }} [options] logging interval in seconds. 0 will stop interval logging.
+     */
+    public async chanperf(options?: { interval: number }): Promise<SceneGraphCommandResponse> {
+        let command = 'chanperf';
+
+        if (options) {
+            // TODO: revisit this as channelId support is documented but the command does not seem to work. Device returns 'ERR: unknown arg: <channelId>'
+            // command = options?.channelId ? `${command} ${options.channelId}` : command;
+            command = options?.interval > -1 ? `${command} -r ${options.interval}` : command;
+        }
+
+        return this.exec(command);
+    }
+
+    /**
+     * Clear all caches that can affect channel launch time.
+     */
+    public async clearLaunchCaches(): Promise<SceneGraphCommandResponse> {
+        return this.exec('clear_launch_caches');
+    }
+
+    /**
+     * Displays frames-per-second and free memory on-screen. Leverage this tool to optimize your channel UI. It presents a 1-second moving average of the current frame rate.
+     * @param {('off'|'on'|'toggle')} option
+     */
+    public async fpsDisplay(option: 'off'|'on'|'toggle'): Promise<SceneGraphCommandResponse> {
+        let command = 'fps_display';
+
+        if (option !== 'toggle') {
+            command = `${command} ${option === 'on' ? 1 : 0}`;
+        }
+
+        let response = await this.exec(command);
+        if (!response?.error) {
```

Comment:
This is a little weird. !response means we got no response. Then !response?.error could either mean "got no response" or "got a response but there is no error". It might be worth handling the "got no response" in a separate if statement. Or, if we are guaranteed to get a response, then remove the `?.`

Reply:
We are always going to get a response. I'll remove the NOT condition. This is just me not playing with the ?. enough.